apps/arweave/src/ar_mining_worker.erl

```diff
@@ -542,100 +553,204 @@ handle_task({compute_h2_for_peer, Candidate, _ExtraArgs}, State) ->
 			process_chunks(
-				WhichChunk, Candidate, RangeStart, Nonce + SubChunksPerChunk, SubChunksPerChunk,
-				SubChunksPerRecallRange, [{ChunkEndOffset, Chunk} | ChunkOffsets], SubChunkSize, Count, State1
+				WhichChunk, Candidate, RangeStart, Nonce + NoncesPerChunk, NoncesPerChunk,
+				NoncesPerRecallRange, [{ChunkEndOffset, Chunk} | ChunkOffsets], SubChunkSize, Count, State1
 			);
-		{true, false, chunk2} ->
-			%% Skip this nonce.
+		{true, _, chunk2} ->
+			%% Skip these nonces (starting from Nonce to Nonce + NoncesPerChunk - 1).
 			%% Nonce falls in a chunk which wasn't read from disk (for example, because there are holes
 			%% in the recall range), e.g. the nonce is in the middle of a non-existent chunk.
 			%% Mark single chunk2 as missing or remove it if the corresponding chunk is already read and H1 is calculated.
+			?LOG_ERROR([{event, mining_worker_process_chunks_nonce_offset_less_than_chunk2_start_offset},
+				{chunk, WhichChunk},
+				{nonce, Nonce},
+				{nonce_offset, NonceOffset},
+				{chunk_start_offset, ChunkStartOffset},
+				{chunk_end_offset, ChunkEndOffset},
+				{chunk_offsets_left, length(ChunkOffsets)},
+				{partition, State#state.partition_number}]),
 			State1 = mark_single_chunk2_missing_or_drop(Nonce, Candidate, State),
 			process_chunks(
-				WhichChunk, Candidate, RangeStart, Nonce + SubChunksPerChunk, SubChunksPerChunk,
-				SubChunksPerRecallRange, [{ChunkEndOffset, Chunk} | ChunkOffsets], SubChunkSize, Count, State1
+				WhichChunk, Candidate, RangeStart, Nonce + NoncesPerChunk, NoncesPerChunk,
+				NoncesPerRecallRange, [{ChunkEndOffset, Chunk} | ChunkOffsets], SubChunkSize, Count, State1
 			);
-		{false, true, _} ->
+		{_, true, _} ->
 			%% Skip this chunk.
 			%% Nonce falls in a chunk beyond the current chunk offset, (for example, because we
 			%% read extra chunk in the beginning of recall range). Move ahead to the next
 			%% chunk offset.
 			%% No need to remove anything from cache, as the nonce is still in the recall range.
+			%%
+			%% Does this ever happen?
+			?LOG_ERROR([{event, mining_worker_process_chunks_nonce_offset_greater_than_chunk_end_offset},
+				{chunk, WhichChunk},
+				{nonce, Nonce},
+				{nonce_offset, NonceOffset},
+				{chunk_start_offset, ChunkStartOffset},
+				{chunk_end_offset, ChunkEndOffset},
+				{chunk_offsets_left, length(ChunkOffsets)},
+				{partition, State#state.partition_number}]),
 			process_chunks(
-				WhichChunk, Candidate, RangeStart, Nonce, SubChunksPerChunk,
-				SubChunksPerRecallRange, ChunkOffsets, SubChunkSize, Count, State
+				WhichChunk, Candidate, RangeStart, Nonce, NoncesPerChunk,
+				NoncesPerRecallRange, ChunkOffsets, SubChunkSize, Count, State
 			);
 		{false, false, _} ->
 			%% Process all sub-chunks in Chunk, and then advance to the next chunk.
-			State1 = process_all_sub_chunks(WhichChunk, Chunk, Candidate, Nonce, State),
+			State1 = process_all_sub_chunks(WhichChunk, Chunk, 0, Candidate, Nonce, State),
```

Comment:
Why do we start from 0 here rather than Nonce? (I've gotten my self all twisted - legit question)

Reply:
This one was only used for debugging purposes, I wanted to ensure we're always getting 32 subchunks here (which was the case). I'll remove this counter, it is not needed.

Reply:
Now I get your frustration, in the end it should be 31, not 32. But I'm not seeing any logs from line 730 on the server.